### PR TITLE
Support Dart 3 Record types

### PR DESCRIPTION
**Pull request by kevmoo**
Closes https://github.com/google/json_serializable.dart/issues/1222

### CODE REVIEW

**Comment by kevmoo** [comment]
The implementation is less than 200 lines. Most of this is auto-generated test code.

File: json_serializable/lib/src/type_helpers/record_helper.dart [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    bool defaultProvided,
+  ) {
+    if (targetType is! RecordType) return null;
+
+    final items = <Object>[];
+
+    const paramName = r'$jsonValue';
+
+    var index = 1;
+    for (var field in targetType.positionalFields) {
```
Comment (jakemac53) [inline json_serializable/lib/src/type_helpers/record_helper.dart:24]: It feels a bit weird to serialize the positional fields as a map with the `$*` names. Although it does simplify things some.

I could also see serializing them as a List, with positional fields followed by a nested Map for the named fields, so like (1, 2, {name: 'Foo'}) would be like `[1, 2, {"name": "Foo"}]`.

Ultimately I don't know how much it matters how we serialize records, but that would be a bit more true to the original format?

**Review by jakemac53 — APPROVED** [review]

File: json_serializable/lib/src/type_helpers/record_helper.dart (hunk 1 above)
Comment (kevmoo) [reply json_serializable/lib/src/type_helpers/record_helper.dart:24]: I don't think we gain anything by encoding to look like the definition syntax. It makes encode and decode more complex. I'm trying to do the thing that's just easiest.

**Comment by Dampfwalze** [comment]
Are there examples or other documentation on how to serialize and deserialize records?

**Pull request merged**